plugins/modules/web_infrastructure/jira.py

```diff
@@ -662,35 +407,263 @@ def main():
+        self.vars.meta = self.post(url, self.vars.username, self.vars.password, self.vars.timeout, data)
+
+    @cause_changes(on_success=True)
+    def operation_link(self):
+        data = {
+            'type': {'name': self.vars.linktype},
+            'inwardIssue': {'key': self.vars.inwardissue},
+            'outwardIssue': {'key': self.vars.outwardissue},
+        }
+        url = self.vars.restbase + '/issueLink/'
+        self.vars.meta = self.post(url, self.vars.username, self.vars.password, self.vars.timeout, data)
+
+    def attach(self):
+        v = self.vars
+        filename = v.attachment.get('filename')
+        content = v.attachment.get('content')
+
+        if not any((filename, content)):
+            raise ValueError('at least one of filename or content must be provided')
+        mime = v.attachment.get('mimetype')
+
+        if not os.path.isfile(filename):
+            raise ValueError('The provided filename does not exist: %s' % filename)
+
+        content_type, data = self._prepare_attachment(filename, content, mime)
+
+        url = v.restbase + '/issue/' + v.issue + '/attachments'
+        return True, self.post(
+            url, v.username, v.password, v.timeout, data, content_type=content_type,
+            additional_headers={"X-Atlassian-Token": "no-check"}
+        )
+
+    # Ideally we'd just use prepare_multipart from ansible.module_utils.urls, but
+    # unfortunately it does not support specifying the encoding and also defaults to
+    # base64. Jira doesn't support base64 encoded attachments (and is therefore not
+    # spec compliant. Go figure). I originally wrote this function as an almost
+    # exact copypasta of prepare_multipart, but ran into some encoding issues when
+    # using the noop encoder. Hand rolling the entire message body seemed to work
+    # out much better.
+    #
+    # https://community.atlassian.com/t5/Jira-questions/Jira-dosen-t-decode-base64-attachment-request-REST-API/qaq-p/916427
+    #
+    # content is expected to be a base64 encoded string since Ansible doesn't
+    # support passing raw bytes objects.
+    @staticmethod
+    def _prepare_attachment(filename, content=None, mime_type=None):
+        def escape_quotes(s):
+            return s.replace('"', '\\"')
+
+        boundary = "".join(random.choice(string.digits + string.ascii_letters) for dummy in range(30))
+        name = to_native(os.path.basename(filename))
+
+        if not mime_type:
+            try:
+                mime_type = mimetypes.guess_type(filename or '', strict=False)[0] or 'application/octet-stream'
+            except Exception:
+                mime_type = 'application/octet-stream'
+        main_type, sep, sub_type = mime_type.partition('/')
+
+        if not content and filename:
+            with open(to_bytes(filename, errors='surrogate_or_strict'), 'rb') as f:
+                content = f.read()
+        else:
+            try:
+                content = base64.decode(content)
```

Comment:
Hi @DWSR , Pycham is moaning about this line:
- apparently `base64.decode()` returns `None` so the content used in line 585 below is null
- `base64.decode()` expects two parameters

Could you please double check on this?

Reply:
Are you just testing if the content *is* in base64 or do you really need to decode it? I would expect the former, but asking to be on the safe side. If indeed it's the first case, then maybe what you need is:
```suggestion
                dummy = base64.b64decode(content)
```

Reply:
It needs to be decoded. The `content` param for the module is expected to be a base64 string in order to handle situations where the file content itself is binary (which Ansible can't handle).

Not sure how I missed this either. 🤦

Reply:
It would be great if this could be moved to a separate PR so it can be backported to stable-2.

Reply:
Or, well, there should probably be a separate PR on stable-2 fixing that for stable-2 once this one is merged which is mostly unrelated to this PR, except parts of the changelog fragment appear in it and this change.